artiq/dashboard/datasets.py

```diff
@@ -81,6 +82,63 @@ def get_edit_widget(self, initial_value):
     def get_edit_widget_value(self):
         return self.edit_widget.text()
 
+class Creator(QtWidgets.QDialog):
+    def __init__(self, parent, dataset_ctl):
+        QtWidgets.QDialog.__init__(self, parent=parent)
+        self.dataset_ctl = dataset_ctl
+
+        self.setWindowTitle("Create dataset")
+        grid = QtWidgets.QGridLayout()
+        self.setLayout(grid)
+
+        grid.addWidget(QtWidgets.QLabel("Name:"), 0, 0)
+        grid.addWidget(self.get_key_widget(), 0, 1)
+
+        grid.addWidget(QtWidgets.QLabel("Value:"), 1, 0)
+        grid.addWidget(self.get_edit_widget(), 1, 1)
+        self.data_type = QtWidgets.QLabel("data type")
+        grid.addWidget(self.data_type, 1, 2)
+        self.edit_widget.textChanged.connect(self.dtype)
+
+        buttons = QtWidgets.QDialogButtonBox(
+            QtWidgets.QDialogButtonBox.Ok | QtWidgets.QDialogButtonBox.Cancel)
+        grid.setRowStretch(2, 1)
+        grid.addWidget(buttons, 3, 0, 1, 2)
+        buttons.accepted.connect(self.accept)
+        buttons.rejected.connect(self.reject)
+
+    def accept(self):
+        try:
+            key = self.get_create_widget_text()
+            value = self.get_create_widget_value()
+            asyncio.ensure_future(self.dataset_ctl.set(key, pyon.decode(value)))
+            QtWidgets.QDialog.accept(self)
+        except:
+            logger.error("Cannot create dataset: "
+                                 "type is not supported",)
+            return
+
+    def get_key_widget(self):
+        self.key_widget = QtWidgets.QLineEdit()
```

Comment:
I find a `get_*` method that overwrites a member to be rather confusing. I'd either just use `self.key_widget` throughout, or make the `get_*` methods lazily create the line edits (i.e. something like `if self.key_widget is None: … return self.key_widget`).